### PR TITLE
Avoid Matching on RTPS Relay

### PR DESCRIPTION
**Pull request by iguessthislldo**
Also remove `GuidPrefixEqual` since nothing was using it as a struct and replaced it with a normal function.

### CODE REVIEW

File: dds/DCPS/DiscoveryBase.h [hunk 1]
```diff
@@ -1216,10 +1220,6 @@ namespace OpenDDS {
           return; // Possible and ok, since lock is released
         }
 
-        if (!writer_local && !reader_local) {
-          return;
-        }
-
```
Comment (mitza-oci) [inline dds/DCPS/DiscoveryBase.h:1222]: Should this be removed? When there is at least one local endpoint on this topic, should the function continue past this point when both sides are non-local?

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/DiscoveryBase.h (hunk 1 above)
Comment (iguessthislldo) [reply dds/DCPS/DiscoveryBase.h:1222]: Then what about restoring my previous change in `match_endpoints` where we return before checking non-locals?

Comment (mitza-oci) [reply dds/DCPS/DiscoveryBase.h:1222]: Any change that helps scalability and doesn't hurt correctness is fair game

**Review by mitza-oci — APPROVED** [review]

**Review by jrw972 — APPROVED** [review]

File: dds/DCPS/DiscoveryBase.h [hunk 2]
```diff
@@ -1009,6 +1009,10 @@ namespace OpenDDS {
           discovered_endpoints = td.discovered_subscriptions();
         }
 
+        if (local_endpoints.empty()) {
+          return;
+        }
```
Comment (iguessthislldo) [inline dds/DCPS/DiscoveryBase.h:1014]: It seems that when this function is called `local_endpoints` can be empty and `repoId` can be a local endpoint that has to make it to the `discovered_endpoints` for loop or else it causes problems.

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**